solidity/contracts/isms/hook/PolygonZkevmIsm.sol

```diff
@@ -0,0 +1,178 @@
+    ) external view override {
+        bytes memory messageId = abi.encodePacked(_message.id());
+        revert OffchainLookup(
+            address(this),
+            offchainUrls,
+            messageId,
+            PolygonZkevmIsm.verify.selector,
+            _message
+        );
+    }
+
+    /**
+     * @dev Calls the Polygon zkEVM bridge to claim the message.
+     * @param _metadata from CCIP call
+     * @return A boolean indicating whether the message was successfully verified and processed.
+     */
+    function verify(
+        bytes calldata _metadata,
+        bytes calldata
+    )
+        external
+        override(AbstractMessageIdAuthorizedIsm, IInterchainSecurityModule)
+        returns (bool)
+    {
+        (
+            bytes32[32] memory smtProof,
+            uint32 index,
+            bytes32 mainnetExitRoot,
+            bytes32 rollupExitRoot,
+            uint32 originNetwork,
+            address originAddress,
+            uint32 destinationNetwork,
+            address destinationAddress,
+            uint256 amount,
+            bytes memory payload
+        ) = abi.decode(
+                _metadata,
+                (
+                    bytes32[32],
+                    uint32,
+                    bytes32,
+                    bytes32,
+                    uint32,
+                    address,
+                    uint32,
+                    address,
+                    uint256,
+                    bytes
+                )
+            );
+
+        zkEvmBridge.claimMessage(
+            smtProof,
+            index,
+            mainnetExitRoot,
+            rollupExitRoot,
+            originNetwork,
+            originAddress,
+            destinationNetwork,
+            destinationAddress,
+            amount,
+            payload
+        );
+
+        return true;
+    }
+
+    /**
+     * @dev Callback function for Zkevm bridge.
+     * Verifies the received message.
+     * @inheritdoc IBridgeMessageReceiver
+     */
+    function onMessageReceived(
+        address,
+        uint32,
+        bytes memory data
+    ) external payable override {
+        require(data.length == 32, "PolygonZkevmIsm: data must be 32 bytes");
+        bytes32 messageId = abi.decode(data, (bytes32));
+        require(
+            _isAuthorized(),
+            "AbstractMessageIdAuthorizedIsm: sender is not the hook"
+        );
+        require(
+            msg.value < 2 ** VERIFIED_MASK_INDEX,
+            "AbstractMessageIdAuthorizedIsm: msg.value must be less than 2^255"
+        );
+
+        verifiedMessages[messageId] = msg.value.setBit(VERIFIED_MASK_INDEX);
+        emit ReceivedMessage(messageId);
+    }
+
+    /**
+     * @dev Checks if the origin chain message sender is the hook address.
+     * @inheritdoc AbstractMessageIdAuthorizedIsm
+     */
+    function _isAuthorized() internal view override returns (bool) {
+        bytes32 originSender = abi.decode(msg.data[4:], (bytes32));
```

Comment:
I dont really understand this

Reply:
you need to compare the `address originSender` from `onMessageReceived` to the `authorizedHook`

Reply:
Umm okay, So I wanted to use the `_isAuthorized()` for the verification, tho now I feel like just using a normal require check for `originSender == TypeCasts.bytes32ToAddress(authorizedHook)` would be better.

I was just slicing the function selector and just decoding the address to match it with authorisedHook